Fix the OnNavigatedTo event and the OnAppearing event to return the ScrollToAsync method call

cat0363 opened the pull request:
This PR fixes the problem that the call does not return when the ScrollToAsync method is 
called at the timing of OnNavigatedTo, OnAppearing events.

### Description of Change

If the ScrollToAsync method is called in the OnNavigatedTo or OnAppearing event, the Handler is null 
at this point, so I added code to handle the scroll request in the Loaded event.

[src\Controls\src\Core\ScrollView\ScrollView.cs]

    void OnScrollToRequested(ScrollToRequestedEventArgs e)
    {
        CheckTaskCompletionSource();
        ScrollToRequested?.Invoke(this, e);

        if (Handler is not null)
        {
            Handler.Invoke(nameof(IScrollView.RequestScrollTo), ConvertRequestMode(e).ToRequest());
        }
        else
        {
            Loaded += (sender, args) =>
            {
                Dispatcher.Dispatch(() =>
                {
                    Handler?.Invoke(nameof(IScrollView.RequestScrollTo), ConvertRequestMode(e).ToRequest());
                });
            };
        }
    }

On Windows, the ViewChanged event doesn't fire until the ScrollView's position is updated, so I added code to call the ScrollFinish method if the requested scroll position is the same as the current scroll position when processing a scroll request.

[src\Core\src\Handlers\ScrollView\ScrollViewHandler.Windows.cs]

    public static void MapRequestScrollTo(IScrollViewHandler handler, IScrollView scrollView, object? args)
    {
        if (args is ScrollToRequest request)
        {
            if (handler.PlatformView.HorizontalOffset == request.HorizontalOffset &&
                handler.PlatformView.VerticalOffset == request.VerticalOffset)
            {
                handler.VirtualView.ScrollFinished();
            }
            else
            {
                handler.PlatformView.ChangeView(request.HorizontalOffset, request.VerticalOffset, null, request.Instant);
            }
        }
    }

<!-- Enter description of the fix in this section -->

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes #15387 

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

I used the following code for testing.

https://github.com/cat0363/Maui-IssueScrollToAsync2.git

The following log is output to the Editor before and after calling the ScrollToAsync method.

        txtLog.Text += ("Before ScrollToAsync" + Environment.NewLine);
        await svTestItem.ScrollToAsync(0, 200, false);
        txtLog.Text += ("After ScrollToAsync" + Environment.NewLine);

Below is the execution result.

[iOS]

<table>
  <tr>
    <td>[OnNavigatedTo]</td>
    <td>[OnAppearing]</td>
  </tr>
  <tr>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/b187cdd5-ccf2-4def-97b0-3e3f26724b45" /></td>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/b1e835b9-b9b4-4a64-a3d6-99b26628671d" /></td>
  </tr>
</table>

[Android]

<table>
  <tr>
    <td>[OnNavigatedTo]</td>
    <td>[OnAppearing]</td>
  </tr>
  <tr>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/711776f7-751a-47fd-8c0b-013c741781f9" /></td>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/debbb7c3-bbc0-4c67-9448-e62869510a85" /></td>
  </tr>
</table>

[Windows]

<table>
  <tr>
    <td>[OnNavigatedTo]</td>
    <td>[OnAppearing]</td>
  </tr>
  <tr>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/83c919e6-f30f-4995-96e4-b8f29a4c40e4" /></td>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/497d9dc8-68c6-4a28-9852-a8a3fe29fbcf" /></td>
  </tr>
</table>

You can see that the ScrollToAsync method call is returned in the OnNavigatedTo event and OnAppearing event on all platforms.
The log is already displayed when the screen is started, but it is the log when the ScrollToAsync method is called in the OnNavigateTo event and OnAppearing event that is called when the screen is displayed for the first time.

cat0363 committed:
Fix the OnNavigatedTo event and the OnAppearing event to return the ScrollToAsync method call

# Conflicts:
#	src/Controls/src/Core/ScrollView/ScrollView.cs

jfversluis commented:
/azp run MAUI-UITests-public

cat0363 committed:
Add unit tests

cat0363 commented:
I wrote the unit test code.
This is my first time writing test code that uses NavigationPage, so if there are any mistakes, could you point them out?

rmarinho commented:
@cat0363 can you get the latest changes from main please? I think this could have been fixed by https://github.com/dotnet/maui/commit/5af60d34225e3101bfca8cd7d834ce9e10e42bf7

cat0363 commented:
@rmarinho , Thank you for letting me know.
I got the latest for main and verified it.

Below are the verification results.

[iOS]

<table>
  <tr>
    <td>[OnNavigatedTo]</td>
    <td>[OnAppearing]</td>
  </tr>
  <tr>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/cce73f9b-3c86-4be2-90dd-b79d646e85f9" /></td>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/8e48ad2d-bfc6-48b1-b57c-6d5e5c3f6180" /></td>
  </tr>
</table>

On iOS, calling the ScrollToAsync method in the OnNavigatedTo and OnAppearing events when the screen is first displayed does not work as intended. 

[Android]

<table>
  <tr>
    <td>[OnNavigatedTo]</td>
    <td>[OnAppearing]</td>
  </tr>
  <tr>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/d13c83d7-de26-4b6f-93bc-acd640cadd38" /></td>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/f5554a6c-0428-46e4-8747-45b949ba7782" /></td>
  </tr>
</table>

On Android, calling the ScrollToAsync method in the OnNavigatedTo and OnAppearing events when the screen is first displayed does not work as intended. 

[Windows]

<table>
  <tr>
    <td>[OnNavigatedTo]</td>
    <td>[OnAppearing]</td>
  </tr>
  <tr>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/30af6051-ef14-4b29-8b46-335cdf6458eb
" /></td>
    <td><video controls width="300" src="https://github.com/dotnet/maui/assets/125236133/a9ac3e81-b8c8-4869-bac5-8b9865a3cd50" /></td>
  </tr>
</table>

On Windows, calling the ScrollToAsync method in the OnNavigatedTo and OnAppearing events when the screen is first displayed does not work as intended. Calling the ScrollToAsync method in the OnAppearing event when transitioning from another screen does not work as intended.

All in all, the result was not what I expected.

The PR I created shows that the ScrollToAsync method works as intended even when the screen is displayed for the first time.
I expected the ScrollToAsync method to work as intended when the screen first appeared, and not when it was navigated to from another screen.

cat0363 commented:
With 5af60d3, the problem is still there. Is there any other solution?
A stale tag was added, but the problem is still there.

cat0363 commented:
Even with .NET 8, the problem continues to occur.
If you wait synchronously for the ScrollToAsync method inside the OnNavigatedTo event, it will remain blocked.

cat0363 commented:
@rmarinho  , As far as I investigated, the method call does not finish only if the ScrollToAsync method is called when the ScrollView's size is not determined. 
Calling the ScrollToAsync method when the height and width of the ScrollView are 0 and not determined will fail.
I hope this will give you a hint on how to fix it.

rmarinho commented:
Can you rebase on main I will pull and take a look

cat0363 commented:
@rmarinho , I found a solution to the problem and uploaded the modified code.

The latest source code for main is designed so that if the ScrollToAsync method is called before the Handler is finalized, the scroll process will be performed after the Handler is finalized using TaskCompletionSource.
However, the scrolling occurs after the Handler is confirmed, but the TrySetResult method of TaskCompletionSource is not called, so the task is not completed and remains in a waiting state.

For this reason, I modified it as follows.

[src\Controls\src\Core\ScrollView\ScrollView.cs]

    private protected override void OnHandlerChangedCore()
    {
        base.OnHandlerChangedCore();

        if (Handler is not null && _pendingScrollToRequested is not null)
        {
            OnScrollToRequested(_pendingScrollToRequested);
            _pendingScrollToRequested = null;
            _scrollCompletionSource.TrySetResult(true);
        }
    }

In addition to the above, if the ScrollToAsync method is called before the Handler is finalized, the CheckTaskCompletionSource method will be called twice in total, so we fixed the TaskCompletionSource generation process in the CheckTaskCompletionSource method.

[src\Controls\src\Core\ScrollView\ScrollView.cs]

    void CheckTaskCompletionSource()
    {
        if (_scrollCompletionSource != null && _scrollCompletionSource.Task.Status == TaskStatus.Running)
        {
            _scrollCompletionSource.TrySetCanceled();
        }
        if (_pendingScrollToRequested == null)
        {
            _scrollCompletionSource = new TaskCompletionSource<bool>();
        }
    }

By modifying it as above, if you call the ScrollToAsync method within the OnNavigatedTo, OnAppearing event, it will no longer wait forever.

cat0363 commented:
Modify the previous code and upload it again.

cat0363 commented:
I made some changes today, but I'll revert to the latest version of main.
After applying the latest version of main, both Android and iOS no longer wait forever after calling the ScrollToAsync method, but on Android, the screen does not scroll correctly the first time it is displayed. On Android, the scroll position remains at (0,0).

It's probably the same problem as the one below.
https://github.com/dotnet/maui/issues/14594

cat0363 commented:
The latest code verification results for main are listed below.

On Windows, if the call to the ScrollToAsync method does not change the position of the ScrollBar, the call to the ScrollToAsync method does not finish. Therefore, in addition to the latest source code for main, you will need:

[src\Core\src\Handlers\ScrollView\ScrollViewHandler.Windows.cs]

```
public static void MapRequestScrollTo(IScrollViewHandler handler, IScrollView scrollView, object? args)
{
    if (args is ScrollToRequest request)
    {
        if (handler.PlatformView.HorizontalOffset == request.HorizontalOffset &&
            handler.PlatformView.VerticalOffset == request.VerticalOffset)
        {
            handler.VirtualView.ScrollFinished();
        }
        else
        {
            handler.PlatformView.ChangeView(request.HorizontalOffset, request.VerticalOffset, null, request.Instant);
        }
    }
}
```

On iOS, calling the ScrollToAsync method within the OnNavigatedTo, OnAppearing events works as intended.

On Android, calling the ScrollToAsync method within the OnNavigatedTo, OnAppearing event does not work as intended only on first appearance.

PureWeen commented:
/azp run

jfversluis commented:
/azp run MAUI-public

PureWeen commented:
/rebase
